api/python/quilt3/backends/base.py

```diff
@@ -0,0 +1,174 @@
+        return self.base.is_local()
+
+    @property
+    @abc.abstractmethod
+    def root_path(self):
+        pass
+
+    @property
+    @abc.abstractmethod
+    def pointers_global_dir(self) -> PhysicalKey:
+        pass
+
+    @abc.abstractmethod
+    def pointers_dir(self, pkg_name: str) -> PhysicalKey:
+        pass
+
+    @abc.abstractmethod
+    def pointer_pk(self, pkg_name: str, pointer_name: str) -> PhysicalKey:
+        pass
+
+    @abc.abstractmethod
+    def pointer_latest_pk(self, pkg_name: str) -> PhysicalKey:
+        pass
+
+    @property
+    @abc.abstractmethod
+    def manifests_global_dir(self) -> PhysicalKey:
+        pass
+
+    @abc.abstractmethod
+    def manifests_package_dir(self, pkg_name) -> PhysicalKey:
+        pass
+
+    @abc.abstractmethod
+    def manifest_pk(self, pkg_name: str, top_hash: str) -> PhysicalKey:
+        pass
+
+    @abc.abstractmethod
+    def list_packages(self):
+        pass
+
+    @abc.abstractmethod
+    def list_package_versions(self, pkg_name: str):
+        pass
+
+    @abc.abstractmethod
+    def delete_package(self, pkg_name: str, top_hash: str = None):
+        pass
+
+    @abc.abstractmethod
+    def push_manifest(self, pkg_name: str, top_hash: str, manifest_data: bytes):
+        pass
+
+    @abc.abstractmethod
+    def resolve_top_hash(self, pkg_name: str, hash_prefix: str) -> str:
+        pass
+
+    @abc.abstractmethod
+    def shorten_top_hash(self, pkg_name: str, top_hash: str) -> str:
+        pass
+
+
+class PackageRegistryV1(PackageRegistry):
+    root_path = '.quilt'
+
+    @property
+    def pointers_global_dir(self) -> PhysicalKey:
+        return self.root.join('named_packages/')
+
+    def pointers_dir(self, pkg_name: str) -> PhysicalKey:
+        return self.root.join(f'named_packages/{pkg_name}/')
+
+    def pointer_pk(self, pkg_name: str, pointer_name: str) -> PhysicalKey:
+        return self.root.join(f'named_packages/{pkg_name}/{pointer_name}')
+
+    def pointer_latest_pk(self, pkg_name: str) -> PhysicalKey:
+        return self.pointer_pk(pkg_name, self.latest_tag_name)
+
+    @property
+    def manifests_global_dir(self) -> PhysicalKey:
+        return self.root.join('packages/')
+
+    def manifests_package_dir(self, pkg_name) -> PhysicalKey:
+        return self.manifests_global_dir  # TODO: does this make sense?
+
+    def manifest_pk(self, pkg_name: str, top_hash: str) -> PhysicalKey:
+        return self.root.join(f'packages/{top_hash}')
+
+    def delete_package(self, pkg_name: str, top_hash: str = None):
+        package_path = self.pointers_dir(pkg_name)
+        paths = list(list_url(package_path))
+        if not paths:
+            raise QuiltException("No such package exists in the given directory.")
+
+        if top_hash is not None:
+            top_hash = self.resolve_top_hash(pkg_name, top_hash)
+            deleted = []
+            remaining = []
+            for path, pkg_hash in self.list_package_versions(pkg_name):
```

Comment:
this is dangerous if multiple packages have the same path. this can actually happen, e.g. if someone has an empty package with several names. or, better example, a `foo/dev` gets pushed to `foo/prod` and user deletes `foo/prod`. we can punt on deep_delete until v2 where manifests are not shared across handles.